File: packages/distance/src/token/tversky.ts
/**
 * Tversky index — asymmetric set similarity measure.
 *
 * Reduces to Jaccard when alpha = beta = 1.
 * Reduces to Sorensen-Dice when alpha = beta = 0.5.
 *
 * Time: O(m + n)
 */

import {
  charFrequencyMap,
  intersectCount,
  totalCount,
  CHAR_FREQ_SIZE,
  buildCharFreqArray,
} from "../utils";

const _freqA = new Int32Array(CHAR_FREQ_SIZE);
const _freqB = new Int32Array(CHAR_FREQ_SIZE);

/**
 * Options for Tversky index.
 */
export interface ITverskyOptions {
  /**
   * Weight for elements unique to the first set (a).
   * @default 1
   */
  alpha?: number;

  /**
   * Weight for elements unique to the second set (b).
   * @default 1
   */
  beta?: number;
}

/**
 * Compute the Tversky index between two strings based on character multiset.
 *
 * T(A, B; α, β) = |A ∩ B| / (|A ∩ B| + α|A \ B| + β|B \ A|)
 *
 * @param a - First string
 * @param b - Second string
 * @param options - alpha and beta weights
 * @returns Tversky index in [0, 1]
 */
export function tversky(a: string, b: string, options: ITverskyOptions = {}): number {
  const alpha = options.alpha ?? 1;
  const beta = options.beta ?? 1;

  // ASCII fast path
  _freqA.fill(0);
  _freqB.fill(0);
  if (buildCharFreqArray(_freqA, a) && buildCharFreqArray(_freqB, b)) {
    let intersection = 0;
    let totalA = 0;
    let totalB = 0;
    for (let i = 0; i < CHAR_FREQ_SIZE; i++) {
      const va = _freqA[i];
      const vb = _freqB[i];
      intersection += va < vb ? va : vb;
      totalA += va;
      totalB += vb;
    }
    const onlyA = totalA - intersection;
    const onlyB = totalB - intersection;
    const denominator = intersection + alpha * onlyA + beta * onlyB;
    return denominator === 0 ? 1 : intersection / denominator;
  }

  const freqA = charFrequencyMap(a);
  const freqB = charFrequencyMap(b);

  const intersection = intersectCount(freqA, freqB);
  const totalA = totalCount(freqA);
  const totalB = totalCount(freqB);

  const onlyA = totalA - intersection;
  const onlyB = totalB - intersection;

  const denominator = intersection + alpha * onlyA + beta * onlyB;
  if (denominator === 0) return 1;

  return intersection / denominator;
}

File: packages/distance-wasm/src/token/tversky.rs
/// Tversky index — asymmetric set similarity.
///
/// T(A, B; α, β) = |A ∩ B| / (|A ∩ B| + α|A \ B| + β|B \ A|)
/// Default: α = β = 1 (same as Jaccard)
/// Time: O(m + n)

pub fn tversky(a: &str, b: &str) -> f64 {
    tversky_with_params(a, b, 1.0, 1.0)
}

pub fn tversky_with_params(a: &str, b: &str, alpha: f64, beta: f64) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();

    let mut freq_a = [0u32; 128];
    let mut freq_b = [0u32; 128];

    if !crate::utils::build_char_freq(&mut freq_a, a_bytes) || !crate::utils::build_char_freq(&mut freq_b, b_bytes) {
        return tversky_fallback(a_bytes, b_bytes, alpha, beta);
    }

    let intersection = crate::utils::intersect_count(&freq_a, &freq_b);
    let total_a = crate::utils::total_count(&freq_a);
    let total_b = crate::utils::total_count(&freq_b);

    let only_a = total_a - intersection;
    let only_b = total_b - intersection;
    let denominator = intersection as f64 + alpha * only_a as f64 + beta * only_b as f64;
    if denominator == 0.0 { 1.0 } else { intersection as f64 / denominator }
}

fn tversky_fallback(a: &[u8], b: &[u8], alpha: f64, beta: f64) -> f64 {
    let mut fa = std::collections::HashMap::<u8, u32>::new();
    let mut fb = std::collections::HashMap::<u8, u32>::new();
    for &c in a { *fa.entry(c).or_insert(0) += 1; }
    for &c in b { *fb.entry(c).or_insert(0) += 1; }

    let (smaller, larger) = if fa.len() <= fb.len() { (&fa, &fb) } else { (&fb, &fa) };
    let mut intersection = 0u32;
    for (k, &v) in smaller {
        if let Some(&w) = larger.get(k) { intersection += v.min(w); }
    }
    let total_a: u32 = fa.values().sum();
    let total_b: u32 = fb.values().sum();

    let only_a = total_a - intersection;
    let only_b = total_b - intersection;
    let denominator = intersection as f64 + alpha * only_a as f64 + beta * only_b as f64;
    if denominator == 0.0 { 1.0 } else { intersection as f64 / denominator }
}
